osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -155,61 +155,107 @@ public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user
         /// Duplicates <see cref="CurrentItem"/> into the database.
         /// </summary>
         /// <param name="db">The database connection.</param>
-        private async Task duplicateCurrentItem(IDatabaseAccess db)
+        private async Task duplicateCurrentItem(IDatabaseAccess db) => await addItem(db, new MultiplayerPlaylistItem
         {
-            var newItem = new MultiplayerPlaylistItem
-            {
-                OwnerID = CurrentItem.OwnerID,
-                BeatmapID = CurrentItem.BeatmapID,
-                BeatmapChecksum = CurrentItem.BeatmapChecksum,
-                RulesetID = CurrentItem.RulesetID,
-                AllowedMods = CurrentItem.AllowedMods,
-                RequiredMods = CurrentItem.RequiredMods
-            };
-
-            newItem.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, newItem));
-            room.Playlist.Add(newItem);
-
-            await hub.OnPlaylistItemAdded(room, newItem);
+            OwnerID = CurrentItem.OwnerID,
+            BeatmapID = CurrentItem.BeatmapID,
+            BeatmapChecksum = CurrentItem.BeatmapChecksum,
+            RulesetID = CurrentItem.RulesetID,
+            AllowedMods = CurrentItem.AllowedMods,
+            RequiredMods = CurrentItem.RequiredMods
+        });
+
+        private async Task addItem(IDatabaseAccess db, MultiplayerPlaylistItem item)
+        {
+            // Add the item to the list with an initial "infinite" ID, in order to notify updatePlaylistItem()
+            // that the item should be de-prioritised and events should not be propagated for it.
+            item.ID = long.MaxValue;
+            room.Playlist.Add(item);
+            await updatePlaylistOrder(db);
+
+            // Actually add the item to the database and invoke callback on the hub.
+            item.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));
+            await hub.OnPlaylistItemAdded(room, item);
         }
 
         /// <summary>
         /// Updates <see cref="CurrentItem"/> and the playlist item ID stored in the room's settings.
         /// </summary>
         private async Task updateCurrentItem()
         {
-            MultiplayerPlaylistItem newItem;
+            // Pick the next non-expired playlist item by playlist order, or default to the most-recently-expired item.
+            MultiplayerPlaylistItem nextItem = room.Playlist.Where(i => !i.Expired).OrderBy(i => i.PlaylistOrder).FirstOrDefault()
```

Comment:
To confirm, the database ordering is only valid for non-expired items, and will overlap with expired? So it's basically a throwaway value which is meaningless (should explicitly never be read) once the item is expired?

Wondering if the `expired` column should be removed and this one should be used instead, setting it to `-1` for expired.

Reply:
Yeah it's pretty much an undefined value for expired items. I originally had a doc on it but the semantics have changed a few times and the doc is now gone.